### ISSUE TITLE
External authentication (i.e. OpenID Connect)

### ISSUE THREAD

**Issue by alexanderadam** [letsblockit/letsblockit#169]
### Describe the bug

It's surely a luxury feature that might only be relevant once [there's an official Docker container](https://github.com/letsblockit/letsblockit/issues/69) available.
But it would be nice to have OIDC support once people can host it on their own.

This way people could use their external auth services like [Authelia](https://www.authelia.com/), [Keycloak](https://www.keycloak.org/), [Authentik](https://goauthentik.io/) or [GitLab](https://docs.gitlab.com/ee/integration/openid_connect_provider.html).

For selfhosting having the possibility to set the credentials as `ENV` variables would probably be sufficient.

There are also battle tested packages available for Go (_i.e. [`go-oidc`](https://github.com/coreos/go-oidc), [`oidc`](https://github.com/ericchiang/oidc) and [others](https://pkg.go.dev/search?q=openid+client&m=package)_).

**PS:** the letsblockit issue tracker should probably also have a feature category 😉 

### Your browser and device

-

### Reproduction steps

```bash
1. Sign in via external provider
```

**Comment by xvello** [comment]
Hello Alexander!

The server is currently coupled with [Ory Kratos](https://github.com/ory/kratos) that can be self-hosted, although its setup is non-trivial. I actually chose them because oauth2 is a mess to implement (complex and insecure) and they designed a much saner flow. I like [their detailed blog post on the topic](https://www.ory.sh/oauth2-openid-connect-do-you-need-use-cases-examples/).

That said, I have been working towards decoupling the auth layer into an interface, to allow other providers to be added. The second provider I would implement would be a simple [HTTP Basic auth](https://developer.mozilla.org/en-US/docs/Web/HTTP/Authentication), sourcing password hashes from the database.

Once that work is done, it would be pretty trivial to write an auth provider that integrates [oauth2-proxy](https://oauth2-proxy.github.io/oauth2-proxy/). I will probably need to change the database `user_id` from UUID though, as the common ID for oauth2 is the user email.

I feel that it's better to outsource the oauth2 work to that proxy instead of re-implementing it again, plus some people might already be running it for other applications. What do you think?

**Comment by alexanderadam** [comment]
Bonjour Xavier,

> Once that work is done, it would be pretty trivial to write an auth provider that integrates [oauth2-proxy](https://oauth2-proxy.github.io/oauth2-proxy/). I will probably need to change the database `user_id` from UUID though, as the common ID for oauth2 is the user email.

I'm not quite sure how this works to be honest. So OAuth2 Proxy is another application that does OIDC/OAuth2 on top of LetsBlockIt then?
This means users having an existing IDP would spin up a proxy container that sits between the reverse proxy and LetsBlockIt instead of the IDP directly.
When the user authenticates, it will be redirected to IDP which will redirect to the proxy and then to LetsBlockIt with the right headers(?).

In that case this would probably even be fine _without_ the necessity of having OAuth2 Proxy since [Authelia](https://www.authelia.com/docs/deployment/supported-proxies/#how-can-the-backend-be-aware-of-the-authenticated-users), [Keycloak](https://brianturchyn.net/traefik-forwardauth-support-with-keycloak/), [Authentik](https://goauthentik.io/docs/providers/proxy/forward_auth) and likely others support forward auth.
It's probably just important that the naming is consistent and working.

**Comment by xvello** [comment]
Indeed, oauth2-proxy intercepts the connections, enforces auth, and proxies it down to the server with an added HTTP header holding the user's email. We could indeed support any IDP solution, as long as it:

  - allows skipping auth for `/list/*`, the list download endpoint
  - passes one unique property of the user (ID, username or email) as a HTTP header. We can then make that header configurable in the server config

HTTP basic could even be implemented this way, as even nginx supports it. I'm not sure nginx injects the username as a header though.

**Comment by xvello** [comment]
Thanks for raising awareness on the project, it's very helpful :bow: 

The `pkg/server` package has become a big ball of mud, so it's due for some cleanups and decoupling. The auth system is as good a candidate as any to start.

**Comment by xvello** [comment]
Automatically closed that issue with my commit, I'm opening https://github.com/letsblockit/letsblockit/issues/201 to track the documentation efforts.
